README.md
# dahdi-linux-complete-3.0.0-3.0.0

Supported Hardware
Digital Cards

- wcte43x:
  * Digium TE435: PCI express quad-port T1/E1/J1
  * Digium TE436: PCI quad-port T1/E1/J1
  * Digium TE235: PCI express dual-port T1/E1/J1
  * Digium TE236: PCI dual-port T1/E1/J1
- wcte13xp:
  * Digium TE131: PCI express single-port T1/E1/J1
  * Digium TE133: PCI express single-port T1/E1/J1 with echocan
  * Digium TE132: PCI single-port T1/E1/J1
  * Digium TE134: PCI single-port T1/E1/J1 with echocan
- wct4xxp:
  * OpenVox D230E/P: PCI-Express or PCI dual-port T1/E1/J1
  * OpenVox D430E/P: PCI-Express or PCI quad-port T1/E1/J1
  * OpenVox D830E/P: PCI-Express or PCI eight-port T1/E1/J1
  * Digium TE205P/TE207P/TE210P/TE212P: PCI dual-port T1/E1/J1
  * Digium TE405P/TE407P/TE410P/TE412P: PCI quad-port T1/E1/J1
  * Digium TE220: PCI-Express dual-port T1/E1/J1
  * Digium TE420: PCI-Express quad-port T1/E1/J1
  * Digium TE820: PCI-Express eight-port T1/E1/J1
- opvxd115:
  * OpenVox D130E/P: PCI-Express or PCI single-port T1/E1/J1
- wcb4xxp:
  * OpenVox B400E/P: PCI-Express or PCI quad-port BRI
  * OpenVox B200E/P: PCI-Express or PCI dual-port BRI
  * OpenVox BE400E/P: PCI-Express or PCI quad-port BRI
  * OpenVox BE200E/P: PCI-Express or PCI dual-port BRI
  * Digium B410: PCI quad-port BRI
  * Digium B233: PCI-Express dual-port BRI with echo can
  * Digium B234: PCI dual-port dual-port BRI with echo can
  * Digium B433: PCI-Express quad-port BRI with echo can
  * Digium B434: PCI quad-port BRI with echo can
- zaphfc:
   * OpenVox B100E/P: PCI-Express or PCI single-port BRI


Analog Cards
~~~~~~~~~~~~
- wctdm:
  * OpenVox A400E/P: PCI-Express or PCI 4 analog ports
- opvxa1200:
  * OpenVox A800E/P: PCI-Express or PCI 8 analog ports
  * OpenVox A1200E/P: PCI-Express or PCI 12 analog ports
- opvxa24xx:
  * OpenVox A810E/P: PCI-Express or PCI 8 analog ports
  * OpenVox A1610E/P: PCI-Express or PCI 16 analog ports
  * OpenVox A2410E/P: PCI-Express or PCI 24 analog ports
- wcaxx:
  * Digium A8A: PCI up to 8 mixed FXS/FXO ports
  * Digium A8B: PCI express up to 8 mixed FXS/FXO ports
  * Digium A4A: PCI up to 4 mixed FXS/FXO ports
  * Digium A4B: PCI express up to 4 mixed FXS/FXO ports
- wctdm24xxp:
  * Digium TDM2400P/AEX2400: up to 24 analog ports
  * Digium Hx8 Series: Up to 8 analog or BRI ports
- xpp: Xorcom Astribank: a USB connected unit of up to 32 ports
  (including the digital BRI and E1/T1 modules)
  
Hybird Cards
~~~~~~~~~~~~
- opvxx200
  * OpenVox X204E/P: PCI-Express or PCI with 4 modules
                     FXO200M -- 2 FXO ports
                     FXS200M -- 2 FXS ports
                     D100M -- single E1/T1/J1 port
                     B200M -- 2 Ports BRI (TE/NT)


ChangeLog
2014-01-24 Russ Meyerriecks <rmeyerriecks@digium.com>

	* dahdi-linux-complete version 2.9.0-rc1+2.9.0-rc1 released.

2013-12-19 Russ Meyerriecks <rmeyerriecks@digium.com>
	
	* dahdi-linux-complete version 2.8.0.1+2.8.0 released.

2013-08-21 Russ Meyerriecks <rmeyerriecks@digium.com>
	
	* dahdi-linux-complete version 2.7.0.1+2.7.0.1 released.

2013-05-29 Russ Meyerriecks <rmeyerriecks@digium.com>
	
	* dahdi-linux-complete version 2.7.0-rc1+2.7.0-rc1 released.

2013-03-13 Russ Meyerriecks <rmeyerriecks@digium.com>

	* dahdi-linux-complete version 2.6.3-rc1+2.6.3-rc1 released.

2013-02-21 Russ Meyerriecks <rmeyerriecks@digium.com>

	* dahdi-linux-complete version 2.6.2+2.6.2 released.

2012-04-18 Shaun Ruffell <sruffell@digium.com>

        * dahdi-linux-complete version 2.6.0+2.6.0 released.

2012-01-03 Shaun Ruffell <sruffell@digium.com>

        * dahdi-linux-complete version 2.6.0+2.6.0 released.

2011-11-02 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.6.0-rc1+2.6.0-rc1 released.

2011-08-05 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.5.0+2.5.0 released.

2011-04-11 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.4.1.2+2.4.1 released.

2011-03-31 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.4.1.1+2.4.1 released.

2011-03-03 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.4.1+2.4.1 released.

2010-08-31 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.4.0+2.3.0 released.

2010-05-17 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.3.0.1+2.3.0 released.

2010-04-12 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.3.0+2.3.0 released.

2010-01-23 Russ Meyerriecks <rmeyerriecks@digium.com>
	
	* dahdi-linux-complete version 2.2.1.1+2.2.1.1 released.

2010-01-11 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.2.1+2.2.1 released.

2009-12-15 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.2.1-rc2+2.2.1-rc2 released.

2009-06-30 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.2.0.1+2.2.0 released.

2009-06-23 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.2.0+2.2.0 released.

2009-05-27 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.2.0-rc5+2.2.0-rc3 released.

2009-05-11 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.2.0-rc4+2.2.0-rc2 released.

2009-05-07 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.2.0-rc3+2.2.0-rc2 released.

2009-04-29 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.2.0-rc2+2.2.0-rc2 released.

2009-02-02 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.2.0-rc1+2.2.0-rc1 released.

2008-12-28 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.1.0.4+2.1.0.2 released.

2008-12-17 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.1.0.3+2.1.0.2 released.

2008-12-15 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.1.0.2+2.1.0.2 released.

2008-12-11 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.1.0.1+2.1.0.1 released.

2008-12-09 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.1.0+2.1.0 released.

2008-11-25 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.1.0-rc5+2.1.0-rc5 released.

2008-11-17 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.1.0-rc4+2.1.0-rc4 released.

2008-11-10 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.1.0-rc3+2.1.0-rc3 released.

2008-11-05 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.1.0-rc2+2.1.0-rc2 released.

2008-11-03 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.1.0-rc1+2.1.0-rc1 released.

2008-09-29 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.0.0+2.0.0 released.

2009-09-18 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.0.0-rc6+2.0.0-rc3 released.

2008-09-17 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux-complete version 2.0.0-rc5+2.0.0-rc3 released.

2008-09-08 Russell Bryant <russell@digium.com>

	* dahdi-linux-complete version 2.0.0-rc4+2.0.0-rc2 released.

2008-08-20 Kevin P. Fleming <kpfleming@digium.com>

	* dahdi-linux-complete version 2.0.0-rc3+2.0.0-rc2 released.

2008-08-06 Kevin P. Fleming <kpfleming@digium.com>

	* dahdi-linux-complete version 2.0.0-rc2+2.0.0-rc1 released.

2008-08-06 Kevin P. Fleming <kpfleming@digium.com>

	* dahdi-linux-complete version 2.0.0-rc1+2.0.0-rc1 released.



linux/drivers/dahdi/opvxa24xx/a24xx.c
/*
 * OpenVox A24xx FXS/FXO Interface Driver for Zapata Telephony interface
 *
 * Written by MiaoLin<miaolin@openvox.cn>
 * $Id: a24xx.c 185 2010-12-14 07:58:51Z yangshugang $
 
 * Copyright (C) 2005-2010 OpenVox Communication Co. Ltd,
 *
 * All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */
 
#include <linux/spinlock.h>
#include <linux/jiffies.h>

#include <dahdi/kernel.h>

#include "base.h"
#include "proslic.h"

#define BIT_EC_PRESENT      (1<<0)

#define VPM_DEFAULT_DTMFTHRESHOLD 1000

/* indirect_resg */
static alpha  indirect_regs[] =
{
{0,255,"DTMF_ROW_0_PEAK",0x55C2},
{1,255,"DTMF_ROW_1_PEAK",0x51E6},
{2,255,"DTMF_ROW2_PEAK",0x4B85},
{3,255,"DTMF_ROW3_PEAK",0x4937},
{4,255,"DTMF_COL1_PEAK",0x3333},
{5,255,"DTMF_FWD_TWIST",0x0202},
{6,255,"DTMF_RVS_TWIST",0x0202},
{7,255,"DTMF_ROW_RATIO_TRES",0x0198},
{8,255,"DTMF_COL_RATIO_TRES",0x0198},
{9,255,"DTMF_ROW_2ND_ARM",0x0611},
{10,255,"DTMF_COL_2ND_ARM",0x0202},
{11,255,"DTMF_PWR_MIN_TRES",0x00E5},
{12,255,"DTMF_OT_LIM_TRES",0x0A1C},
{13,0,"OSC1_COEF",0x7B30},
{14,1,"OSC1X",0x0063},
{15,2,"OSC1Y",0x0000},
{16,3,"OSC2_COEF",0x7870},
{17,4,"OSC2X",0x007D},
{18,5,"OSC2Y",0x0000},
{19,6,"RING_V_OFF",0x0000},
{20,7,"RING_OSC",0x7EF0},
{21,8,"RING_X",0x0160},
{22,9,"RING_Y",0x0000},
{23,255,"PULSE_ENVEL",0x2000},
{24,255,"PULSE_X",0x2000},
{25,255,"PULSE_Y",0x0000},
//{26,13,"RECV_DIGITAL_GAIN",0x4000},	// playback volume set lower
{26,13,"RECV_DIGITAL_GAIN",0x2000},	// playback volume set lower
{27,14,"XMIT_DIGITAL_GAIN",0x4000},
//{27,14,"XMIT_DIGITAL_GAIN",0x2000},
{28,15,"LOOP_CLOSE_TRES",0x1000},
{29,16,"RING_TRIP_TRES",0x3600},
{30,17,"COMMON_MIN_TRES",0x1000},
{31,18,"COMMON_MAX_TRES",0x0200},
{32,19,"PWR_ALARM_Q1Q2",0x0ff4},
{33,20,"PWR_ALARM_Q3Q4",0x6e7e},
{34,21,"PWR_ALARM_Q5Q6",0x0ff4},
{35,22,"LOOP_CLOSURE_FILTER",0x8000},
{36,23,"RING_TRIP_FILTER",0x0320},
{37,24,"TERM_LP_POLE_Q1Q2",0x0012},
{38,25,"TERM_LP_POLE_Q3Q4",0x0012},
{39,26,"TERM_LP_POLE_Q5Q6",0x0012},
{40,27,"CM_BIAS_RINGING",0x0C00},
{41,64,"DCDC_MIN_V",0x0C00},
{42,255,"DCDC_XTRA",0x1000},
{43,66,"LOOP_CLOSE_TRES_LOW",0x1000},
};

void __a24xx_wait_just_a_bit(int foo)
{
	long newjiffies;
	newjiffies = jiffies + foo;
	while(jiffies < newjiffies);
}

void __a24xx_setcard(void *wc_dev, int card)
{
	struct a24xx_dev *dev = (struct a24xx_dev *)(wc_dev);
	if (dev->curcard != card) {
		__opvx_a24xx_setcard(dev->mem32, card);
		dev->curcard = card;
	}
}

inline void __a24xx_spi_setreg(struct a24xx_dev *wc_dev, int card, unsigned char reg, unsigned char value)
{
	__opvx_a24xx_spi_setreg((void *)wc_dev, wc_dev->mem32, card, wc_dev->modtype[card], reg, value, __a24xx_setcard);
}

inline unsigned char __a24xx_spi_getreg(struct a24xx_dev *wc_dev, int card, unsigned char reg)
{
	return __opvx_a24xx_spi_getreg((void *)wc_dev, wc_dev->mem32, card, wc_dev->modtype[card], reg, __a24xx_setcard);
}

int __a24xx_malloc_chunk(struct a24xx_dev *wc_dev,unsigned int frq) 
{
	__opvx_a24xx_set_chunk(&(wc_dev->readchunk), &(wc_dev->writechunk),frq);
	wc_dev->readdma = wc_dev->writedma + frq * DAHDI_MAX_CHUNKSIZE * (MAX_NUM_CARDS) * 2;

	return 0;
}

static unsigned char __translate_3215(unsigned char address)
{
	int x;
	for (x=0;x<sizeof(indirect_regs)/sizeof(indirect_regs[0]);x++) {
		if (indirect_regs[x].address == address) {
			address = indirect_regs[x].altaddr;
			break;
		}
	}
	return address;
}

int __a24xx_wait_access(struct a24xx_dev *wc_dev, int card)
{
	unsigned char data = 0;
	long origjiffies;
	int count = 0;

	#define MAX 6000 /* attempts */

	origjiffies = jiffies;
	/* Wait for indirect access */
	while (count++ < MAX) {
		data = __a24xx_spi_getreg(wc_dev, card, I_STATUS);
		if (!data) {
			return 0;
		}
	}

	if(count > (MAX-1)) {
		printk(" ##### Loop error (%02x) #####\n", data);
	}

	return 0;
}

int __a24xx_proslic_setreg_indirect(struct a24xx_dev *wc_dev, int card, unsigned char address, unsigned short data)
{
	int res = -1;
	/* Translate 3215 addresses */
	if (wc_dev->flags[card] & FLAG_3215) {
		address = __translate_3215(address);
		if (address == 255) {
			return 0;
		}
	}
	if(!__a24xx_wait_access(wc_dev, card)) {
		__a24xx_spi_setreg(wc_dev, card, IDA_LO,(unsigned char)(data & 0xFF));
		__a24xx_spi_setreg(wc_dev, card, IDA_HI,(unsigned char)((data & 0xFF00)>>8));
		__a24xx_spi_setreg(wc_dev, card, IAA,address);
		res = 0;
	}
	
	return res;
}

int __a24xx_proslic_getreg_indirect(struct a24xx_dev *wc_dev, int card, unsigned char address)
{
	int res = -1;
	char *p=NULL;
	/* Translate 3215 addresses */
	if (wc_dev->flags[card] & FLAG_3215) {
		address = __translate_3215(address);
		if (address == 255) {
			return 0;
		}
	}
	if (!__a24xx_wait_access(wc_dev, card)) {
		__a24xx_spi_setreg(wc_dev, card, IAA, address);
		if (!__a24xx_wait_access(wc_dev, card)) {
			unsigned char data1, data2;
			data1 = __a24xx_spi_getreg(wc_dev, card, IDA_LO);
			data2 = __a24xx_spi_getreg(wc_dev, card, IDA_HI);
			res = data1 | (data2 << 8);
		} else {
			p = "Failed to wait inside\n";
		}
	} else {
		p = "failed to wait\n";
	}
	if (p) {
		printk("%s\n", p);
	}
	return res;
}

void __a24xx_vpm_setpresent(struct a24xx_dev *wc_dev)
{
	wc_dev->vpm = BIT_EC_PRESENT;
    if  ( (wc_dev->fwversion&0xffff) > 0x3)
	    __opvx_a24xx_vpm_setpresent_v2(wc_dev->mem32);
	else 
	    __opvx_a24xx_vpm_setpresent(wc_dev->mem32);
	printk("OpenVox VPM: Present and operational servicing %d span(s)\n", 1/*wc_dev->numspans*/);
}

void a24xx_spi_setreg(struct a24xx_dev *wc_dev, int card, unsigned char reg, unsigned char value)
{
	unsigned long flags;
	spin_lock_irqsave(&wc_dev->lock, flags);
	__a24xx_spi_setreg(wc_dev, card, reg, value);
	spin_unlock_irqrestore(&wc_dev->lock, flags);
}

unsigned char a24xx_spi_getreg(struct a24xx_dev *wc_dev, int card, unsigned char reg)
{
	unsigned long flags;
	unsigned char ret;
	spin_lock_irqsave(&wc_dev->lock, flags);
	ret = __a24xx_spi_getreg(wc_dev, card, reg);
	spin_unlock_irqrestore(&wc_dev->lock, flags);
	return ret;
}

static inline unsigned int a24xx_oct_in(struct a24xx_dev *wc_dev, const unsigned int addr)
{
	unsigned long flags;
	unsigned int ret;

	spin_lock_irqsave(&wc_dev->lock, flags);
    if  ( (wc_dev->fwversion&0xffff) > 0x3)
	    ret = __opvx_a24xx_oct_in_v2(wc_dev->mem32, addr);
	else 
	    ret = __opvx_a24xx_oct_in(wc_dev->mem32, addr);
	spin_unlock_irqrestore(&wc_dev->lock, flags);
	
	return ret;
}

static inline void a24xx_oct_out(struct a24xx_dev *wc_dev, const unsigned int addr, const unsigned int value)
{
	unsigned long flags;

	spin_lock_irqsave(&wc_dev->lock, flags);
    if  ( (wc_dev->fwversion&0xffff) > 0x3)
	    __opvx_a24xx_oct_out_v2(wc_dev->mem32, addr, value);
	else 
	    __opvx_a24xx_oct_out(wc_dev->mem32, addr, value);
	spin_unlock_irqrestore(&wc_dev->lock, flags);
}

void oct_set_reg(void *data, unsigned int reg, unsigned int val)
{
	struct a24xx_dev *wc_dev = data;
	a24xx_oct_out(wc_dev, reg, val);
}

unsigned int oct_get_reg(void *data, unsigned int reg)
{
	struct a24xx_dev *wc_dev = data;
	unsigned int ret;
	ret = a24xx_oct_in(wc_dev, reg);
	return ret;
}

void a24xx_reset_spi(struct a24xx_dev *wc_dev, int card)
{
	unsigned long flags;
	spin_lock_irqsave(&wc_dev->lock, flags);
	__opvx_a24xx_reset_spi(wc_dev, card, __a24xx_setcard);
	spin_unlock_irqrestore(&wc_dev->lock, flags);
}

